query-projector/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#[macro_use]
+extern crate error_chain;
+extern crate rusqlite;
+
+extern crate mentat_core;
+extern crate mentat_db;                 // For value conversion.
+extern crate mentat_query;
+extern crate mentat_query_algebrizer;
+extern crate mentat_query_sql;
+extern crate mentat_sql;
+
+use std::iter;
+
+use rusqlite::{
+    Row,
+    Rows,
+};
+
+use mentat_core::{
+    TypedValue,
+};
+
+use mentat_db::{
+    SQLValueType,
+    TypedSQLValue,
+};
+
+use mentat_query::{
+    Element,
+    FindSpec,
+    PlainSymbol,
+    Variable,
+};
+
+use mentat_query_algebrizer::{
+    AlgebraicQuery,
+    DatomsColumn,
+    QualifiedAlias,
+    /*
+    ConjoiningClauses,
+    DatomsTable,
+    SourceAlias,
+    */
+};
+
+use mentat_query_sql::{
+    ColumnOrExpression,
+    /*
+    Constraint,
+    FromClause,
+    */
+    Name,
+    Projection,
+    ProjectedColumn,
+    /*
+    SelectQuery,
+    TableList,
+    */
+};
+
+error_chain! {
+    types {
+        Error, ErrorKind, ResultExt, Result;
+    }
+
+    foreign_links {
+        Rusqlite(rusqlite::Error);
+    }
+
+    links {
+        DbError(mentat_db::Error, mentat_db::ErrorKind);
+    }
+}
+
+#[derive(Debug)]
+pub enum QueryResults {
+    Scalar(Option<TypedValue>),
+    Tuple(Option<Vec<TypedValue>>),
+    Coll(Vec<TypedValue>),
+    Rel(Vec<Vec<TypedValue>>),
+}
+
+impl QueryResults {
+    pub fn len(&self) -> usize {
+        use QueryResults::*;
+        match self {
+            &Scalar(ref o) => if o.is_some() { 1 } else { 0 },
+            &Tuple(ref o)  => if o.is_some() { 1 } else { 0 },
+            &Coll(ref v)   => v.len(),
+            &Rel(ref v)    => v.len(),
+        }
+    }
+}
+
+type Index = i32;            // See rusqlite::RowIndex.
+type ValueTypeTag = i32;
+enum TypedIndex {
+    Known(Index, ValueTypeTag),
+    Unknown(Index, Index),
+}
+
+impl TypedIndex {
+    /// Look up this index and type(index) pair in the provided row.
+    /// This function will panic if:
+    ///
+    /// - This is an `Unknown` and the retrieved type code isn't an i32.
```

Comment:
Because we construct our SQL projection list, the code that stored the data, and this consumer, a panic here implies that we have a bad bug — we put data of a very wrong type in a row, and thus can't coerce to `Value`, we're retrieving from the wrong place, or our generated SQL is junk.

Generally I'd rather a bug like this cause a crash with a stack rather than propagate up to application error handling, which — let's face it — is normally just "Oops, an error occurred", or showing an empty list, and muffling the problem altogether.